vita3k/motion/src/motion_input.cpp

```diff
@@ -17,23 +22,34 @@ void MotionInput::SetPID(SceFloat new_kp, SceFloat new_ki, SceFloat new_kd) {
 
 void MotionInput::SetAcceleration(const Util::Vec3f &acceleration) {
     accel = acceleration;
+
+    accel.x = std::clamp(accel.x, -AccelMaxValue, AccelMaxValue);
+    accel.y = std::clamp(accel.y, -AccelMaxValue, AccelMaxValue);
+    accel.z = std::clamp(accel.z, -AccelMaxValue, AccelMaxValue);
 }
 
 void MotionInput::SetGyroscope(const Util::Vec3f &gyroscope) {
+    gyro = gyroscope;
+
     if (bias_enabled) {
-        gyro = gyroscope - gyro_bias;
-    } else {
-        gyro = gyroscope;
+        gyro -= gyro_bias;
+    }
+
+    if (deadband_enabled && gyro.Length() < gyro_deadband) {
+        gyro = {};
     }
 
+    gyro.x = std::clamp(gyro.x, -GyroMaxValue, GyroMaxValue);
+    gyro.y = std::clamp(gyro.y, -GyroMaxValue, GyroMaxValue);
+    gyro.z = std::clamp(gyro.z, -GyroMaxValue, GyroMaxValue);
+
     // Auto adjust drift to minimize drift
     if (!IsMoving(0.1f)) {
         gyro_bias = (gyro_bias * 0.9999f) + (gyroscope * 0.0001f);
     }
 
-    if (gyro.Length2() < gyro_threshold) {
-        gyro = {};
-    } else {
+    // Enable gyro compensation if gyro is active
+    if (gyro.Length() > gyro_deadband) {
         only_accelerometer = false;
     }
```

Comment:
only_accelerometer is to detect if there is a gyro at all (it is never set back to true), is it correct to compare it against gyro_deadband, which can be set by the game?

Reply:
When only accelerometer is enabled gyro will be emulated from accelerometer values. This is a counter measure when the motion source doesn't have a gyro sensor.

This flag is permanently disabled if any gyro input is detected. Returning to normal behavior. It can be changed to `if (gyro.Length2() > 0) `